usr/lib/common/uri.c

```diff
@@ -0,0 +1,268 @@
+                                     (a)[0].ulValueLen), \
+                           P11_URI_P_UNRES, buf);        \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_ATTRIBUTE_ALL(sep, txt, a, t, buf)        \
+    do {                                                 \
+        if (((a)[0].type == (t)) &&                      \
+            ((a)[0].pValue)) {                           \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_append(buf, txt);                 \
+            p11_encode(START_END((a)[0].pValue,          \
+                                 (a)[0].ulValueLen),     \
+                       buf);                             \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_ATTRIBUTE_CLASS(sep, txt, a, buf)         \
+    do {                                                 \
+        if (((a)[0].type == CKA_CLASS) &&                \
+            ((a)[0].pValue) &&                           \
+            (c2t((a)[0].pValue))) {                      \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_append(buf, txt);                 \
+            p11_buffer_append(buf, c2t((a)[0].pValue));  \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_VERSION(sep, txt, v, buf)                 \
+    do {                                                 \
+        if (((v).major != (CK_BYTE)-1) ||                \
+            ((v).minor != (CK_BYTE)-1)) {                \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_printf(buf, txt "%d.%d",          \
+                      (v).major, (v).minor);             \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+#define FORMAT_ULONG(sep, txt, ul, buf)                  \
+    do {                                                 \
+        if ((ul) != (CK_ULONG)-1) {                      \
+            p11_buffer_append_len(buf, sep, 1);          \
+            p11_buffer_printf(buf, txt "%lu", ul);       \
+            *sep = P11_URI_SEP_ATTR;                     \
+        }                                                \
+    } while(0)
+
+static const char HEX_CHARS_LOWER[] = "0123456789abcdef";
+
+static char *c2t(CK_OBJECT_CLASS_PTR class)
+{
+    switch (*class) {
+    case CKO_CERTIFICATE:
+        return "cert";
+    case CKO_DATA:
+        return "data";
+    case CKO_PRIVATE_KEY:
+        return "private";
+    case CKO_PUBLIC_KEY:
+        return "public";
+    case CKO_SECRET_KEY:
+        return "secret-key";
+    default:
```

Comment:
How about the other CKO's:
#define CKO_HW_FEATURE        0x00000005
#define CKO_DOMAIN_PARAMETERS 0x00000006
#define CKO_PROFILE           0x00000009UL

Reply:
You're right, if we move it to p11utils.c, we should also implement the other CKOs as well. I not covered them in my first implementation, because the RFC 7512 only mentions the implemented ones.